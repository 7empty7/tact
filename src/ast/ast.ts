import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {
    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRef = {
    kind: 'type_ref',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction)[]
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public' } | { type: 'get' };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attribute: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementCall = {
    kind: 'statement_call',
    id: number,
    expression: ASTOpCall | ASTOpCallStatic,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: string[],
    expression: ASTExpression,
    ref: ASTRef
}

//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementCall | ASTSTatementAssign;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementCall | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_call' || src.kind === 'statement_assign';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}