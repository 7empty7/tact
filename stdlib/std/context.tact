struct Context {
    bounced: Bool;
    sender: Address;
    value: Int;
}

@name(__tact_context_get)
native context(): Context;