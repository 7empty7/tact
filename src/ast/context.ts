import { parse } from "../grammar/grammar";
import { ASTNode, ASTType } from "./ast";

export type VariableRef = {
    name: string,
    node: ASTNode
};

export class CompilerContext {

    static fromSources(sources: string[]) {
        let asts = sources.map(source => parse(source));
        let ctx = new CompilerContext();
        for (let a of asts) {
            for (let e of a.entries) {
                if (e.kind === 'def_struct') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'def_contract') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'primitive') {
                    ctx = ctx.addASTType(e);
                }
            }
        }
        return ctx;
    }

    readonly astTypes: { [key: string]: ASTType };
    readonly variables: { [key: string]: VariableRef };
    readonly shared: { [key: symbol]: any } = {};

    constructor(astTypes: { [key: string]: ASTType } = {}, variables: { [key: string]: VariableRef } = {}, shared: { [key: symbol]: any } = {}) {
        this.astTypes = astTypes;
        this.variables = variables;
        this.shared = shared;
        Object.freeze(this.astTypes);
        Object.freeze(this.shared);
        Object.freeze(this.variables);
        Object.freeze(this);
    }

    addASTType = (ref: ASTType) => {
        if (this.astTypes[ref.name]) {
            throw Error('Type already exists');
        }
        return new CompilerContext({ ...this.astTypes, [ref.name]: ref }, this.variables, this.shared);
    }

    addVariable = (ref: VariableRef) => {
        if (this.variables[ref.name]) {
            throw Error('Variable already exists');
        }
        return new CompilerContext(this.astTypes, { ...this.variables, [ref.name]: ref }, this.shared);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext(this.astTypes, this.variables, { ...this.shared, [store]: sh });
    }
}

export function createContextStore<T>() {
    let symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            let m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}