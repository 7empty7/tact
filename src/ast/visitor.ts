import { ASTArgument, ASTBuiltIn, ASTContract, ASTExpression, ASTField, ASTFunction, ASTNode, ASTStatement, ASTStruct } from "./ast";
import { CompilerContext } from "./context";

export type ContextVisitor = {
    visit(ctx: CompilerContext, node: ASTNode): void;
    visitEnd(ctx: CompilerContext, node: ASTNode): void;
}

function visitDefField(ctx: CompilerContext, visitor: ContextVisitor, s: ASTField): CompilerContext {
    visitor.visit(ctx, s);
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitArg(ctx: CompilerContext, visitor: ContextVisitor, s: ASTArgument): CompilerContext {
    visitor.visit(ctx, s);
    visitor.visitEnd(ctx, s);
    return ctx.addVariable({ name: s.name, node: s });
}

function visitExpression(ctx: CompilerContext, visitor: ContextVisitor, s: ASTExpression): CompilerContext {
    visitor.visit(ctx, s);
    if (s.kind === 'op_binary') {
        visitExpression(ctx, visitor, s.left);
        visitExpression(ctx, visitor, s.right);
    } else if (s.kind === 'op_unary') {
        visitExpression(ctx, visitor, s.right);
    } else if (s.kind === 'op_field') {
        visitExpression(ctx, visitor, s.src);
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitStatement(ctx: CompilerContext, visitor: ContextVisitor, s: ASTStatement): CompilerContext {
    visitor.visit(ctx, s);
    if (s.kind === 'let') {

        // Visit right hand side
        visitExpression(ctx, visitor, s.expression);

        // Add variable to context (AFTER visiting expression to avoid cyclic references)
        ctx = ctx.addVariable({ name: s.name, node: s });
    } else if (s.kind === 'return') {
        // Nothing to do
    } else {
        throw Error('Unknown type');
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitDefFunction(ctx: CompilerContext, visitor: ContextVisitor, s: ASTFunction): CompilerContext {
    visitor.visit(ctx, s);

    // Process all args
    for (let a of s.args) {
        ctx = visitArg(ctx, visitor, a);
    }

    // Process all statements
    for (let a of s.statements) {
        ctx = visitStatement(ctx, visitor, a);
    }

    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitStruct(ctx: CompilerContext, visitor: ContextVisitor, s: ASTStruct): CompilerContext {
    visitor.visit(ctx, s);
    for (let f of s.fields) {
        ctx = visitDefField(ctx, visitor, f);
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitContract(ctx: CompilerContext, visitor: ContextVisitor, s: ASTContract): CompilerContext {
    visitor.visit(ctx, s);

    // Add self to context
    ctx = ctx.addVariable({ name: 'self', node: s });

    // Process all declarations
    for (let d of s.declarations) {
        if (d.kind === 'def_field') {
            ctx = visitDefField(ctx, visitor, d);
        } else if (d.kind === 'def_function') {
            ctx = visitDefFunction(ctx, visitor, d);
        } else {
            throw Error('Unknown type');
        }
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitBuiltIn(ctx: CompilerContext, visitor: ContextVisitor, s: ASTBuiltIn): CompilerContext {
    visitor.visit(ctx, s);
    visitor.visitEnd(ctx, s);
    return ctx;
}

export function visit(ctx: CompilerContext, visitor: ContextVisitor) {
    for (let n of Object.keys(ctx.astTypes)) {
        let t = ctx.astTypes[n];
        if (t.kind === 'def_struct') {
            ctx = visitStruct(ctx, visitor, t);
        } else if (t.kind === 'def_contract') {
            ctx = visitContract(ctx, visitor, t);
        } else if (t.kind === 'built-in') {
            ctx = visitBuiltIn(ctx, visitor, t);
        } else {
            throw Error('Unknown type');
        }
    }
}