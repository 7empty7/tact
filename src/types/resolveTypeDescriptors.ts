import { ASTField, ASTFunction, ASTInitFunction, ASTNativeFunction, ASTTypeRef, throwError } from "../ast/ast";
import { CompilerContext, createContextStore } from "../ast/context";
import { FieldDescription, FunctionArgument, FunctionDescription, InitDescription, TypeDescription, TypeRef } from "./types";

let store = createContextStore<TypeDescription>();
let staticFunctionsStore = createContextStore<FunctionDescription>();

export function resolveTypeDescriptors(ctx: CompilerContext) {
    let types: { [key: string]: TypeDescription } = {};
    let staticFunctions: { [key: string]: FunctionDescription } = {};

    // Register types
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];
        if (types[a.name]) {
            throwError(`Type ${a.name} already exists`, a.ref);
        }
        if (a.kind === 'primitive') {
            types[a.name] = {
                kind: 'primitive',
                name: a.name,
                fields: [],
                functions: [],
                init: null
            };
        } else if (a.kind === 'def_contract') {
            types[a.name] = {
                kind: 'contract',
                name: a.name,
                fields: [],
                functions: [],
                init: null
            };
        } else if (a.kind === 'def_struct') {
            types[a.name] = {
                kind: 'struct',
                name: a.name,
                fields: [],
                functions: [],
                init: null
            };
        }
    }

    function resolveTypeRef(src: ASTTypeRef): TypeRef {
        if (!types[src.name]) {
            throwError('Type ' + src.name + ' not found', src.ref);
        }
        if (src.optional) {
            return {
                kind: 'optional',
                inner: {
                    kind: 'direct',
                    name: src.name
                }
            };
        } else {
            return {
                kind: 'direct',
                name: src.name
            }
        }
    }

    function resolveFunctionDescriptor(self: TypeDescription | null, a: ASTFunction | ASTNativeFunction): FunctionDescription {

        // Resolve return
        let returns: TypeRef | null = null;
        if (a.return) {
            returns = resolveTypeRef(a.return);
        }

        // Resolve args
        let args: FunctionArgument[] = [];
        for (let r of a.args) {
            args.push({
                name: r.name,
                type: resolveTypeRef(r.type),
                as: null
            });
        }

        // Register function
        return {
            name: a.name,
            self,
            args,
            returns,
            ast: a,
            isPublic: a.kind === 'def_function' ? a.attribute.some(a => a.type === 'public') : false,
            isGetter: a.kind === 'def_function' ? a.attribute.some(a => a.type === 'get') : false,
        };
    }

    function resolveInitFunction(a: ASTInitFunction): InitDescription {
        let args: FunctionArgument[] = [];
        for (let r of a.args) {
            args.push({
                name: r.name,
                type: resolveTypeRef(r.type),
                as: null
            });
        }
        return {
            args,
            ast: a
        }
    }

    // Resolve static functions
    for (let f in ctx.astFunctionStatic) {
        let a = ctx.astFunctionStatic[f];
        if (staticFunctions[a.name]) {
            throw Error('Function ' + a.name + ' already exists');
        }

        // Register function
        staticFunctions[a.name] = resolveFunctionDescriptor(null, a);
    }

    // Resolve fields
    function resolveField(src: ASTField, index: number): FieldDescription {
        return { name: src.name, type: resolveTypeRef(src.type), index, as: src.as };
    }
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];

        // Contract
        if (a.kind === 'def_contract') {
            for (const f of a.declarations) {
                if (f.kind !== 'def_field') {
                    continue;
                }
                if (types[a.name].fields.find((v) => v.name === f.name)) {
                    throw Error('Field ' + f.name + ' already exists');
                }
                types[a.name].fields.push(resolveField(f, types[a.name].fields.length));
            }
        }

        // Struct
        if (a.kind === 'def_struct') {
            for (let f of a.fields) {
                if (types[a.name].fields.find((v) => v.name === f.name)) {
                    throw Error('Field ' + f.name + ' already exists');
                }
                types[a.name].fields.push(resolveField(f, types[a.name].fields.length));
            }
        }
    }

    // Resolve contract functions
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];
        if (a.kind === 'def_contract') {
            let s = types[a.name];
            for (let d of a.declarations) {
                if (d.kind === 'def_function') {
                    s.functions.push(resolveFunctionDescriptor(s, d));
                }
                if (d.kind === 'def_init_function') {
                    if (s.init) {
                        throw Error('Init function already exists');
                    }
                    s.init = resolveInitFunction(d);
                }
            }
        }
    }

    // Register types in context
    for (let t in types) {
        ctx = store.set(ctx, t, types[t]);
    }
    for (let t in staticFunctions) {
        ctx = staticFunctionsStore.set(ctx, t, staticFunctions[t]);
    }

    return ctx;
}

export function getType(ctx: CompilerContext, name: string): TypeDescription {
    let r = store.get(ctx, name);
    if (!r) {
        throw Error('Type ' + name + ' not found');
    }
    return r;
}

export function getAllTypes(ctx: CompilerContext) {
    return store.all(ctx);
}

export function getStaticFunction(ctx: CompilerContext, name: string): FunctionDescription {
    let r = staticFunctionsStore.get(ctx, name);
    if (!r) {
        throw Error('Static function ' + name + ' not found');
    }
    return r;
}

export function getAllStaticFunctions(ctx: CompilerContext) {
    return staticFunctionsStore.all(ctx);
}