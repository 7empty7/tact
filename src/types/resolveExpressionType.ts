import { ASTExpression, ASTFunction, ASTOpCall, ASTOpCallStatic } from "../ast/ast";
import { CompilerContext, createContextStore } from "../ast/context";
import { getStaticFunction, getType } from "./resolveTypeDescriptors";
import { TypeDescription } from "./TypeDescription";

let store = createContextStore<{ ast: ASTExpression, description: TypeDescription }>();

type VariableCTX = { [key: string]: TypeDescription };

export function getExpType(ctx: CompilerContext, exp: ASTExpression) {
    let t = store.get(ctx, exp.id);
    if (!t) {
        throw Error('Expression ' + exp.id + ' not found');
    }
    return t.description;
}

export function resolveExpressionTypes(ctx: CompilerContext) {

    // Process all types
    function registerExpType(ctx: CompilerContext, exp: ASTExpression, name: string): CompilerContext {
        if (store.get(ctx, exp.id)) {
            throw Error('Expression ' + exp.id + ' already has a type');
        }
        return store.set(ctx, exp.id, { ast: exp, description: getType(ctx, name) });
    }
    function resolveCall(ctx: CompilerContext, vctx: VariableCTX, exp: ASTOpCall): CompilerContext {
        ctx = resolveExpression(ctx, vctx, exp.src);
        let src = getExpType(ctx, exp.src);
        let f = src.functions[exp.name];
        if (!f) {
            throw Error('Function ' + exp.name + ' not found at type ' + src.name);
        }
        for (let e of exp.args) {
            ctx = resolveExpression(ctx, vctx, e);
        }
        return ctx;
    }
    function resolveStaticCall(ctx: CompilerContext, vctx: VariableCTX, exp: ASTOpCallStatic): CompilerContext {
        for (let e of exp.args) {
            ctx = resolveExpression(ctx, vctx, e);
        }
        return ctx;
    }
    function resolveExpression(ctx: CompilerContext, vctx: VariableCTX, exp: ASTExpression): CompilerContext {
        if (exp.kind === 'boolean') {
            return registerExpType(ctx, exp, 'Bool');
        } else if (exp.kind === 'number') {
            return registerExpType(ctx, exp, 'Int');
        } else if (exp.kind === 'op_binary') {

            // Resolve left and right expressions
            ctx = resolveExpression(ctx, vctx, exp.left);
            ctx = resolveExpression(ctx, vctx, exp.right);
            let le = getExpType(ctx, exp.left).name;
            let re = getExpType(ctx, exp.right).name;

            // Check operands
            let tp: string;
            if (exp.op === '-' || exp.op === '+' || exp.op === '*' || exp.op === '/') {
                if (le !== 'Int') {
                    throw Error('Unsupported type: ' + le);
                }
                if (re !== 'Int') {
                    throw Error('Unsupported type: ' + le);
                }
                tp = 'Int';
            } else if (exp.op === '==' || exp.op === '!=' || exp.op === '<' || exp.op === '<=' || exp.op === '>' || exp.op === '>=') {
                if (le !== 'Int') {
                    throw Error('Unsupported type: ' + le);
                }
                if (re !== 'Int') {
                    throw Error('Unsupported type: ' + le);
                }
                tp = 'Bool';
            } else if (exp.op === '&&' || exp.op === '||') {
                if (le !== 'Bool') {
                    throw Error('Unsupported type: ' + le);
                }
                if (re !== 'Bool') {
                    throw Error('Unsupported type: ' + le);
                }
                tp = 'Bool';
            } else {
                throw Error('Unsupported operator: ' + exp.op);
            }

            // Register result
            return registerExpType(ctx, exp, tp);
        } else if (exp.kind === 'op_unary') {

            // Resolve right side
            ctx = resolveExpression(ctx, vctx, exp.right);

            // Check right type dependent on operator
            let rightType = getExpType(ctx, exp.right).name;
            if (exp.op === '-' || exp.op === '+') {
                if (rightType !== 'Int') {
                    throw Error('Type mistmatch');
                }
            } else if (exp.op === '!') {
                if (rightType !== 'Bool') {
                    throw Error('Type mistmatch');
                }
            } else {
                throw Error('Unknown operator');
            }

            // Register result
            return registerExpType(ctx, exp, rightType);
        } else if (exp.kind === 'id') {
            let v = vctx[exp.value];
            if (!v) {
                throw Error('Resolve expression');
            }
            return registerExpType(ctx, exp, v.name);
        } else if (exp.kind === 'op_field') {
            ctx = resolveExpression(ctx, vctx, exp.src);
            let src = getExpType(ctx, exp.src);
            let field = src.fields[exp.name];
            if (!field) {
                throw Error('Field ' + exp.name + ' not found');
            }
            return registerExpType(ctx, exp, field.type.name);
        } else if (exp.kind === 'op_call') {

            // Resolve call with arguments
            ctx = resolveCall(ctx, vctx, exp);

            // Resolve return value
            let src = getExpType(ctx, exp.src);
            let f = src.functions[exp.name]!;
            if (!f.returns) {
                throw Error('Function ' + exp.name + ' does not return a value');
            }
            return registerExpType(ctx, exp, f.returns.name);
        } else if (exp.kind === 'op_static_call') {

            // Resolve call with arguments
            ctx = resolveStaticCall(ctx, vctx, exp);

            // Resolve static
            let f = getStaticFunction(ctx, exp.name);
            if (!f.returns) {
                throw Error('Function ' + exp.name + ' does not return a value');
            }
            return registerExpType(ctx, exp, f.returns.name);
        }
        throw Error('Unknown expression');
    }
    function resolveFunction(ctx: CompilerContext, vctx: VariableCTX, f: ASTFunction) {
        for (let s of f.statements) {
            if (s.kind === 'statement_let') {
                ctx = resolveExpression(ctx, vctx, s.expression);
                vctx[s.name] = getType(ctx, s.type);
            } else if (s.kind === 'statement_return') {
                ctx = resolveExpression(ctx, vctx, s.expression);
            } else if (s.kind === 'statement_call') {
                if (s.expression.kind === 'op_call') {
                    ctx = resolveCall(ctx, vctx, s.expression);
                } else if (s.expression.kind === 'op_static_call') {
                    ctx = resolveStaticCall(ctx, vctx, s.expression);
                } else {
                    throw Error('Unknown expression');
                }
            } else {
                throw Error('Unknown statement');
            }
        }
        return ctx;
    }


    // Process all contracts
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];
        if (a.kind === 'def_contract') {
            for (let f of a.declarations) {
                if (f.kind === 'def_function') {

                    // Function variables
                    let vctx: VariableCTX = {};
                    vctx['self'] = getType(ctx, a.name);
                    for (let arg of f.args) {
                        vctx[arg.name] = getType(ctx, arg.type);
                    }

                    // Process function
                    ctx = resolveFunction(ctx, vctx, f);
                }
            }
        }
    }

    // Process all functions
    for (let t in ctx.astFunctionStatic) {
        let f = ctx.astFunctionStatic[t];

        // Function variables
        let vctx: VariableCTX = {};
        for (let arg of f.args) {
            vctx[arg.name] = getType(ctx, arg.type);
        }

        // Process function
        ctx = resolveFunction(ctx, vctx, f);
    }

    return ctx;
}

export function getAllExpressionTypes(ctx: CompilerContext) {
    let res: [string, string][] = [];
    let a = store.all(ctx);
    for (let e in a) {
        res.push([e, a[e].description.name]);
    }
    return res;
}