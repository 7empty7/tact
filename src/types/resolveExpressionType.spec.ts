import { CompilerContext } from "../ast/context";
import { resolveExpressionTypes } from "./resolveExpressionType";
import { resolveTypeDescriptors } from "./resolveTypeDescriptors";

const source = `
    primitive Void;
    primitive Int;
    primitive Bool;
    
    contract Test {
        var a: Int;
        fun hello(a: Int, b: Int): Int {
            let c: Int = 123 * a + self.a;
            return c;
        }
    }
`;

describe('resolveExpressionType', () => {
    it('should resolve types', () => {
        let ctx = CompilerContext.fromSources([source]);
        ctx = resolveTypeDescriptors(ctx);
        ctx = resolveExpressionTypes(ctx);
    });
});