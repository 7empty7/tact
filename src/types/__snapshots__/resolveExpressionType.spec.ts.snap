// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveExpressionType should resolve expressions for case-0.tact 1`] = `
[
  [
    "x",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "x == 0",
    "Bool",
  ],
]
`;

exports[`resolveExpressionType should resolve expressions for case-1.tact 1`] = `
[
  [
    "123",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "123 + 10",
    "Int",
  ],
  [
    "x",
    "Int",
  ],
  [
    "123 + 10 == x",
    "Bool",
  ],
]
`;

exports[`resolveExpressionType should resolve expressions for case-2.tact 1`] = `
[
  [
    "123",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "123 + 10",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "(123 + 10) / 10",
    "Int",
  ],
  [
    "x",
    "Int",
  ],
  [
    "(123 + 10) / 10 == x",
    "Bool",
  ],
  [
    "y",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "y > 0",
    "Bool",
  ],
  [
    "(123 + 10) / 10 == x && y > 0",
    "Bool",
  ],
  [
    "z",
    "Bool",
  ],
  [
    "(123 + 10) / 10 == x && y > 0 || z",
    "Bool",
  ],
]
`;

exports[`resolveExpressionType should resolve expressions for case-3.tact 1`] = `
[
  [
    "x",
    "Int",
  ],
  [
    "y",
    "Int",
  ],
  [
    "x + y",
    "Int",
  ],
  [
    "b",
    "Int",
  ],
  [
    "1",
    "Int",
  ],
  [
    "b + 1",
    "Int",
  ],
  [
    "b",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "b > 0",
    "Bool",
  ],
  [
    "z",
    "Bool",
  ],
  [
    "b > 0 || z",
    "Bool",
  ],
]
`;

exports[`resolveExpressionType should resolve expressions for case-4.tact 1`] = `
[
  [
    "x",
    "Int",
  ],
  [
    "y",
    "Int",
  ],
  [
    "x + y",
    "Int",
  ],
  [
    "b",
    "Int",
  ],
  [
    "1",
    "Int",
  ],
  [
    "b + 1",
    "Int",
  ],
  [
    "b",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "b > 0",
    "Bool",
  ],
  [
    "z",
    "Bool",
  ],
  [
    "b > 0 || z",
    "Bool",
  ],
  [
    "w",
    "Int?",
  ],
  [
    "w!!",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "w!! > 0",
    "Bool",
  ],
  [
    "b > 0 || z || (w!! > 0)",
    "Bool",
  ],
]
`;

exports[`resolveExpressionType should resolve expressions for case-5.tact 1`] = `
[
  [
    "p",
    "Point",
  ],
  [
    "p.x",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "p.x == 0",
    "Bool",
  ],
  [
    "p",
    "Point",
  ],
  [
    "p.y",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "p.y == 0",
    "Bool",
  ],
  [
    "p.x == 0 && p.y == 0",
    "Bool",
  ],
  [
    "abi",
    "$ABI",
  ],
  [
    "p",
    "Point",
  ],
  [
    "abi.pack_cell(p)",
    "Cell",
  ],
]
`;

exports[`resolveExpressionType should resolve expressions for case-6.tact 1`] = `
[
  [
    "10",
    "Int",
  ],
  [
    "a",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "a > 0",
    "Bool",
  ],
  [
    "10",
    "Int",
  ],
  [
    "a",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "-10",
    "Int",
  ],
  [
    "a < -10",
    "Bool",
  ],
  [
    "1",
    "Int",
  ],
  [
    "-1",
    "Int",
  ],
  [
    "1",
    "Int",
  ],
]
`;
